first version of the quiz.

bukazoid opened the pull request:
* questions are stored in csv file with comma as a separator
* supported questions with free answer and questions with choosing
correct answer
* when there is several options to choose the QuizProcessor shuffles
them to not allow student to reuse known answers. He has to read
questions and answers again
* test and test resource to check how good questions are loaded
* application is assembled in jar with dependencies file, so only single
jar file to deliver

Commit:
first version of the quiz.

* questions are stored in csv file with comma as a separator
* supported questions with free answer and questions with choosing
correct answer
* when there is several options to choose the QuizProcessor shuffles
them to not allow student to reuse known answers. He has to read
questions and answers again
* test and test resource to check how good questions are loaded
* application is assembled in jar with dependencies file, so only single
jar file to deliver

stvort commented on homework1/src/main/java/ru/yakimov/Quiz.java:
```diff
@@ -0,0 +1,51 @@
+				questions.size(), 100f * correct / questions.size()));
+
+		context.close();
+	}
```
Код из main лучше вынести в отдельный сервис/раннер, оставив только получение бина и вызов метода вывода вопросов.

bukazoid replied:
хорошо, сделал

stvort commented on homework1/src/main/java/ru/yakimov/Quiz.java:
```diff
@@ -0,0 +1,51 @@
+
+		shell.printLine();
+		shell.printLine(String.format("Your result is %s out of %s.\n\tYou you have %.2f%% of correct answers", correct,
+				questions.size(), 100f * correct / questions.size()));
```
В первом задании нужно было сделать только вывод вопросов. Без самого тестирования (там эмуляция, но все равно больше чем требовалось) )

bukazoid replied:
сделал 2 реализации, одна их которых делает только вывод вопросов, другая - не (возможное) будущее

stvort commented on homework1/src/main/java/ru/yakimov/services/QuizReader.java:
```diff
@@ -0,0 +1,49 @@
+import ru.yakimov.domain.QuizQuestion;
+
+@Slf4j
+public class QuizReader {
```
В приложении нет ни одного интерфейса, хотя некоторым классам он бы пригодился. В т.ч. текущему т.к. у ридера может быть множество реализаций

bukazoid replied:
сделал интерфейсы

stvort commented on homework1/src/main/java/ru/yakimov/services/QuizReader.java:
```diff
@@ -0,0 +1,49 @@
+@Slf4j
+public class QuizReader {
+
+	final String quizFile;
+	final QuestionGenerator quizGenerator;
```
Лучше `private`, везде, где это можно.

bukazoid replied:
я где-то в глубине души понимаю что надо, но неохота, не вижу особых проблем если private нет. 
Бонусом просто приятнее видеть просто перечисление полей с типами, чем с дополнительными модификаторами.
Можно я пока не буду этого писать?

stvort commented on homework1/src/main/java/ru/yakimov/services/QuestionGenerator.java:
```diff
@@ -0,0 +1,29 @@
+public class QuestionGenerator {
+	final String CSV_SEPARATOR = ",";
+
+	public QuizQuestion generate(String line) {
+		String[] params = line.split(CSV_SEPARATOR);
+		if (params.length < 2) {
+			log.info("not enough data for question, skip line: {}", line);
+			// not enough data to read
+			return null;
+		}
```
Лучше вернуть `Optional ` или кинуть исключение

bukazoid replied:
сделал Optional

stvort commented on homework1/src/main/java/ru/yakimov/services/QuestionGenerator.java:
```diff
@@ -0,0 +1,29 @@
+
+@Slf4j
+public class QuestionGenerator {
+	final String CSV_SEPARATOR = ",";
```
Лучше `private static`

stvort commented on homework1/src/main/java/ru/yakimov/services/QuestionProcessor.java:
```diff
@@ -0,0 +1,48 @@
+		this.shell = shell;
+	}
+
+	public boolean processQuestion(QuizQuestion question) {
```
Метод занимается одновременно выводом вопроса и проверкой ответа. Можно разбить на приватные методы

stvort commented on homework1/src/main/java/ru/yakimov/services/Shell.java:
```diff
@@ -0,0 +1,39 @@
+	/**
+	 * predefined answer till we have proper input system
+	 */
+	final String answer = "1";
```
Это может быть константой

stvort commented on homework1/src/main/java/ru/yakimov/services/Shell.java:
```diff
@@ -0,0 +1,39 @@
+	final String answer = "1";
+
+	public void printLine(String line) {
+		System.out.println(line);
```
Потоки ввода/вывода тоже лучше получить извне

stvort commented on homework1/src/main/java/ru/yakimov/services/Shell.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	public void printLine(String text, int mult) {
+		for (int i = 0; i < mult; i++) {
+			printText(text);
+		}
```
`printLine(text.repeat(mult));`?

bukazoid replied:
переделал, на java7 такого не было )

stvort commented on homework1/src/main/java/ru/yakimov/services/Shell.java:
```diff
@@ -0,0 +1,39 @@
+	 */
+	final String answer = "1";
+
+	public void printLine(String line) {
```
Можно кстати еще предусмотреть вариант с параметрами для `String.format`

bukazoid replied:
да, согласен и сделал, убрались форматы из мест где они просто мешались

stvort commented on homework1/src/main/java/ru/yakimov/services/Shell.java:
```diff
@@ -0,0 +1,39 @@
+package ru.yakimov.services;
+
+public class Shell {
```
Я бы назвал более общим образом. Типа `IOService` для интерфейса и `ConsoleIOService` для реализации. Так можно будет сделать разные реализации под одним интерфейсом без изменения классов, использующих данный сервис

bukazoid replied:
переделал

stvort commented on homework1/src/test/java/ru/yakimov/QuizReaderTest.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.List;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
```
Лучше все же перейти на JUnit5

bukazoid replied:
в следущем задании

stvort commented on homework1/src/test/java/ru/yakimov/QuizReaderTest.java:
```diff
@@ -0,0 +1,33 @@
+	public void readFileTest() {
+		log.info("test: {}", name.getMethodName());
+
+		List<QuizQuestion> questions = new QuizReader("testquiz.csv", new QuestionGenerator()).readQuestions();
```
С генератором уже не совсем юнит-тест. Чтобы был юнит, надо замокать генератор и проверить, что его метод `generate` был вызван нужное количество раз, а в аргументы метода пришли ожидаемые строки из тестового файла с вопросами (`Mockito.verify`). За проверку того, что из строк получаются нормальные вопросы должен отвечать уже тест генератора.

bukazoid replied:
согласен что не совсем юнит. Было предложено сделать попытку написать тест :)
Переделал на тест генератора. 

5ый юнит и мокито в другой раз.

Commit:
interfaces,
implementation versions for several beans
runner - service to do all work
plus minor changes

Commit:
question generator and test

stvort commented on homework1/src/main/java/ru/yakimov/runner/QuizV1.java:
```diff
@@ -0,0 +1,38 @@
+public class QuizV1 implements Quiz {
+	QuizReader quizReader;
+	UserIO userIO;
+	QuestionProcessorV1 processor;
```
Лучше по интерфейсу. Туда конечно придется прописать методы, он сейчас пуст

stvort commented on homework1/src/main/java/ru/yakimov/services/UserIOConsole.java:
```diff
@@ -0,0 +1,36 @@
+
+	@Override
+	public void printLine(String format, Object... params) {
+		System.out.println(String.format(format, (Object[]) params));
```
А не надо в `Object[]` кастовать. Оно и так массив

stvort commented on homework1/src/main/java/ru/yakimov/services/QuestionGenerator.java:
```diff
@@ -0,0 +1,30 @@
+
+@Slf4j
+public class QuestionGenerator {
+	final String CSV_SEPARATOR = ",";
```
И все же без `static` не совсем константа

Commit:
minor fixes
fields - private/final
